src/write_atsd.c

```diff
@@ -0,0 +1,1273 @@
+    sfree(ak->plugin_instance);
+    sfree(ak->type);
+    sfree(ak->type_instance);
+    sfree(ak);
+
+    sfree(av->value);
+    sfree(av);
+  }
+  return same_value;
+}
+
+static int wa_write_messages(const data_set_t *ds, const value_list_t *vl,
+                             struct wa_callback *cb) {
+
+  char plugin_instance[100];
+  int status;
+  if (0 != strcmp(ds->type, vl->type)) {
+    ERROR("write_atsd plugin: DS type does not match "
+          "value list type");
+    return -1;
+  }
+
+  int i;
+  gauge_t *rates;
+
+  rates = uc_get_rate(ds, vl);
+  if (rates == NULL) {
+    return -1;
+  }
+
+  if (ds->ds_num != vl->values_len) {
+    ERROR("plugin_dispatch_values: ds->type = %s: "
+          "(ds->ds_num = %zu) != "
+          "(vl->values_len = %zu)",
+          ds->type, ds->ds_num, vl->values_len);
+  }
+
+  char sendline[1024];
+  char metric_name[512];
+  cdtime_t now;
+
+  char entity[WA_MAX_LENGTH];
+  int ent_len = sizeof(entity);
+
+  check_entity(entity, ent_len, cb->entity, vl->host, cb->short_hostname);
+
+  now = cdtime();
+  if ((now - cb->last_property_time) > WA_PROPERTY_INTERVAL) {
+    cb->last_property_time = now;
+
+    int ret;
+    struct utsname buf;
+    ret = uname(&buf);
+    if (!ret) {
+
+      snprintf(sendline, sizeof(sendline),
+               "property e:%s ms:%" PRIu64
+               " t:collectd-atsd v:host=%s v:OperatingSystem=\"%s\" "
+               "v:Node=\"%s\" v:Kernel_Release_Version=\"%s\" "
+               "v:OS_Version=\"%s\" v:Hardware=\"%s\"\n",
+               entity, CDTIME_T_TO_MS(vl->time), vl->host, buf.sysname,
+               buf.nodename, buf.release, buf.version, buf.machine);
+    } else {
+      snprintf(sendline, sizeof(sendline),
+               "property e:%s ms:%" PRIu64 " t:collectd-atsd v:host=%s\n",
+               entity, CDTIME_T_TO_MS(vl->time), vl->host);
+    }
+
+    status = wa_send_message(sendline, cb);
+    if (status != 0) {
+      sfree(rates);
+      return (status);
+    }
+  }
+
+  char tmp[512];
+  char tv[50];
+  char ret[128];
+  size_t ret_len = sizeof(ret);
+
+  char *str_location;
+  char *c;
+
+  memset(ret, 0, ret_len);
+
+  atsd_key_t *ak = NULL;
+  atsd_value_t *av = NULL;
+
+  int same_value;
+
+  for (i = 0; i < ds->ds_num; i++) {
+
+    if (isnan(rates[i]))
+      continue;
+
+    status = format_value(ret, ret_len, i, ds, vl, rates);
+    if (status != 0) {
+      sfree(rates);
+      return (status);
+    }
+
+    same_value = 1;
+    sstrncpy(metric_name, cb->prefix, sizeof(metric_name));
+
+/*
+ *  Normalize read plugin signatures according to the ATSD scheme of data
+ *  to prepare series commands from Network API.
+ *  https://github.com/axibase/atsd/tree/master/api/network
+ *
+ *	In general use prefix with plugin as metric name.
+ *	For some plugins calculate additional statistics,
+ *	use types in metric name.
+ *	Build tags from plugin and type instances with small changes.
+ */
+    if (strcasecmp(vl->plugin, "cpu") == 0) {
```

Comment:
This huge if-else not something we can merge. This is way too specialized and we can't possibly keep this list up to date. For example, the code below assumes that the "entropy" plugin reports a single metric, which is a very brittle assumption. It's just too much of a maintenance burden.

It appears that atsd is using key value pairs to identify metrics. I recommend to look at what plugins for other such systems do and mimic their behavior, for example [the write_prometheus plugin](https://github.com/collectd/collectd/blob/master/src/write_prometheus.c#L359-L373).

If you *have* to map collectd metrics to predefined atsd metrics, write that mapping to a file and have the plugin load that. It at all feasible, try to avoid that though.

Example syntax for such a file:

```
entropy/entropy  entropy.available
swap/swap-*      memory.swap_%{type_instance}
```

This way these mappings are not fixed at compile time and the user has control over what happens.